src/CosmosLogStorage.ts

```diff
@@ -0,0 +1,220 @@
+            querySpec = {
+                query: `SELECT * FROM c`
+            }
+        }
+        else if (appId && (!packageIds || packageIds.length == 0)) {
+            querySpec = {
+                query: `SELECT * FROM c WHERE c.appId = @appId`,
+                parameters: [
+                    {
+                        name: "@appId",
+                        value: appId
+                    }
+                ]
+            }
+        }
+        else if (!appId && (packageIds && packageIds.length > 0)) {
+            querySpec = {
+                //query: `SELECT * FROM c WHERE c.packageId = @packageId`,LARS
+                query: `SELECT * FROM c WHERE ARRAY_CONTAINS(@packageList, c.packageId)`,
+                parameters: [
+                    {
+                        name: '@packageList',
+                        value: packageIds
+                        //name: "@packageId",
+                        //value: packageId!
+
+                    }
+                ]
+            }
+        }
+        else {
+            querySpec = {
+                query: `SELECT * FROM c WHERE ARRAY_CONTAINS(@packageList, c.packageId) AND c.appId = @appId`,
+                parameters: [
+                    {
+                        name: "@appId",
+                        value: appId!
+                    },
+                    {
+                        name: '@packageList',
+                        value: packageIds!
+                    }
+                ]
+            }
+        }
+
+        const feedOptions: Cosmos.FeedOptions = {
+            maxItemCount: Math.min(pageSize, MAX_PAGE_SIZE),
+            continuation: continuationToken
+        }
+
+        if (this.container) {
+            const feedResponse = await this.container.items.query(querySpec, feedOptions).fetchNext();
+            return {
+                logDialogs: feedResponse.resources,
+                continuationToken: feedResponse.continuation
+            }
+        }
+        throw new Error("Contained undefined")
+    }
+
+    /** Retrieve a log dialog from storage */
+    public async Get(appId: string, logDialogId: string): Promise<CLM.LogDialog | undefined> {
+        if (this.container) {
+            // this.container.
+            const { resource } = await this.container.item(this.GetDialogDocumentId(appId, logDialogId), appId).read()
+            return resource
+        }
+        return undefined
+    }
+
+    /** Delete a log dialog in storage */
+    public async Delete(appId: string, logDialogId: string): Promise<void> {
+        if (this.container) {
+            await this.container.item(this.GetDialogDocumentId(appId, logDialogId), appId).delete()
+            CLDebug.Log(`Deleted ${logDialogId} / ${this.GetDialogDocumentId(appId, logDialogId)}`)
+        }
+    }
+
+    /** Delete all log dialogs in storage */
+    public async DeleteAll() {
+        if (!this.container) {
```

Comment:
All these guard statements that check container are code smell. It shouldn't be possible to create a storage instance that doesn't have a container. In other words if you have a instance you know it can connect to container and will work. Can these deferred runtime checks be avoided?

Reply:
I'm open to suggestions.  It can't be initialized in the constructor as it is created asynchronously, so the declaration must included undefined.  One option would be to use container! as we know it will exists

Reply:
Could you wrap ctor + init in a static factory method to ensure it's always valid ?

Reply:
If we store the credentials on the object, but i'm hesitant to do so

Reply:
Actually, can do that separately.